### PR TITLE
[e2e]: improve cloud-init user-data specs

### PR DESCRIPTION
**Pull request by enp0s3**
Signed-off-by: enp0s3 <ibezukh@redhat.com>

**What this PR does / why we need it**:
it may happen that the spec connects to the vm console too late thus missing the prompted string from the cloud-init executed script.

with this change, the script creates a file on the fs and the spec waits until login appears, then checking the file existence.

**Special notes for the reviewer:**
In some circumstances I didn't use `DescribeTable` because helper functions such as `cd.ContainerFor` can panic and therefore cannot by called from `Entry`
**Release note**:
```release-note
NONE
```

### CODE REVIEW

File: tests/vmi_cloudinit_test.go
Comment (EdDev) [inline tests/vmi_cloudinit_test.go:154]: You could replace all of this with `console.RunCommand` which embeds most of the logic and is quicker to feedback in case the command fails.

File: tests/vmi_cloudinit_test.go
Comment (EdDev) [inline tests/vmi_cloudinit_test.go:139]: This text adds a lot of overhead and I am not clear how it helps, even when debugging.
IMO it is overused, beyond its original purpose.

Funny thing is, that there is already such logging in `RunVMIAndExpectLaunch` itself, but that is not really my point here.

**Review by EdDev — COMMENTED** [review]
Nice trick.

File: tests/vmi_cloudinit_test.go
Comment (enp0s3) [reply tests/vmi_cloudinit_test.go]: In general it should help someone that is not familiar with the code to write down the test steps. But I didn't notice the helper functions include steps, I've removed them.

Comment (EdDev) [reply tests/vmi_cloudinit_test.go]: The test steps should be clear from the code itself, using good function names.
Comments and logging text tends to rot and be irrelevant.

Comment (enp0s3) [reply tests/vmi_cloudinit_test.go]: @EdDev I meant when someone wants to document the steps in systems like HP QC or Polarion

Comment (enp0s3) [reply tests/vmi_cloudinit_test.go]: @EdDev Yes I agree. `By` should be used for complex specs.

File: tests/vmi_cloudinit_test.go
Comment (EdDev) [inline tests/vmi_cloudinit_test.go:636]: I do not understand why you need a helper function now and I dislike the inner assertion (vs returning an error). But if you still keep it, maybe it is worth using the assertion with an offset so in case of failure we will see the right stack/line ref.

Comment (enp0s3) [reply tests/vmi_cloudinit_test.go]: Agree, it is redundant.

**Review by EdDev — APPROVED** [review]

**Comment by enp0s3** [comment]
@EdDev Thank you for your time!

**Comment by enp0s3** [comment]
/cc @dhiller 
This should fix the flakes we have in downstream

File: tests/vmi_cloudinit_test.go
Comment (enp0s3) [reply tests/vmi_cloudinit_test.go]: @dhiller Great catch! it was a mistake. I've fixed it

**Review by fossedihelm — APPROVED** [review]
Thanks

File: tests/vmi_cloudinit_test.go
Comment (jean-edouard) [inline tests/vmi_cloudinit_test.go:289]: Nit: couldn't you use a `DescribeTable` for this?

**Review by jean-edouard — COMMENTED** [review]
Nice catch! Just one nit in-line, feel free to unhold.
/approve
/hold

**Comment by fossedihelm** [comment]
Thanks @enp0s3 
/lgtm
I leave it to @xpivarc and @jean-edouard  to delete the hold to give him the possibility to view

File: tests/vmi_cloudinit_test.go
Comment (fossedihelm) [inline tests/vmi_cloudinit_test.go:285]: This cannot be used directly. `tests.NewRandomVMIWithEphemeralDiskAndUserdata` uses `NewRandomVMI` which at the end makes an Expectation and this cannot be done in container nodes https://onsi.github.io/ginkgo/#no-assertions-in-container-nodes
This is why this failure:
https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/9242/pull-kubevirt-check-unassigned-tests/1628028781794955264
:)

Comment (enp0s3) [reply tests/vmi_cloudinit_test.go:285]: @fossedihelm I tested it locally with `FDescribetable..` that's why I didn't see this error. Will fix it soon

File: tests/vmi_cloudinit_test.go
Comment (enp0s3) [reply tests/vmi_cloudinit_test.go]: `cd.ContainerDiskFor` can panic therefore I cannot call it from `Entry`, and I wanted to avoid using libvmi in that case because there are some issues with the Cirros builder and adding more cloud-init disks.

**Comment by fossedihelm** [comment]
/lgtm
/retest-required
Thanks @enp0s3 for your patience

**Pull request merged**
